orchestrator/ethereum_gravity/src/utils.rs

```diff
@@ -1,256 +1,186 @@
-    Ok(downcast_uint256(real_num).expect("EventNonce nonce overflow! Bridge Halt!"))
+    Ok(downcast_to_u64(event_nonce).expect("EventNonce nonce overflow! Bridge Halt!"))
 }
 
 /// Gets the gravityID
 pub async fn get_gravity_id(
-    contract_address: EthAddress,
-    caller_address: EthAddress,
-    web3: &Web3,
-) -> Result<String, Web3Error> {
-    let payload = encode_call("state_gravityId()", &[]).unwrap();
-    let val = web3
-        .simulate_transaction(contract_address, 0u8.into(), payload, caller_address, None)
-        .await?;
-    let gravity_id = String::from_utf8(val);
-    match gravity_id {
-        Ok(val) => Ok(val),
-        Err(e) => Err(Web3Error::BadResponse(e.to_string())),
+    gravity_contract_address: EthAddress,
+    eth_client: EthClient,
+) -> Result<String, GravityError> {
+    let contract_call = Gravity::new(gravity_contract_address, eth_client.clone())
+        .state_gravity_id()
+        .from(eth_client.address())
+        .value(U256::zero());
+    let gas_cost = get_call_gas_cost(eth_client.clone()).await?;
+    let contract_call = contract_call
+        .gas(gas_cost.gas)
+        .gas_price(gas_cost.gas_price);
+
+    let gravity_id = contract_call.call().await?;
+    let id_as_string = String::from_utf8(gravity_id.to_vec());
+
+    match id_as_string {
+        Ok(id) => Ok(id),
+        Err(err) => Err(GravityError::GravityContractError(format!(
+            "Received invalid utf8 when getting gravity id {:?}: {}",
+            &gravity_id, err
+        ))),
     }
 }
 
-/// Gets the ERC20 symbol, should maybe be upstreamed
-pub async fn get_erc20_symbol(
-    contract_address: EthAddress,
-    caller_address: EthAddress,
-    web3: &Web3,
-) -> Result<String, GravityError> {
-    let payload = encode_call("symbol()", &[]).unwrap();
+/// Retrieve gas price and limit in a similar fashion to web30's simulate_transaction.
+/// These values are intended to be used in conjunction with eth_call rather than
+/// eth_sendtransaction. In ethers this is represented by `call()` on a ContractCall rather
+/// than `send()`. Using `call()` will not send a transaction from the caller account or
+/// spend gas.
+pub async fn get_call_gas_cost(eth_client: EthClient) -> Result<GasCost, GravityError> {
+    const GAS_LIMIT: u128 = 12450000; // the most Hardhat will allow, will work on Geth
+
+    let caller_balance = eth_client.get_balance(eth_client.address(), None).await?;
+    let latest_block = eth_client.get_block(BlockNumber::Latest).await?.unwrap();
+    let gas_price = latest_block.base_fee_per_gas.unwrap_or(1u8.into()); // "or" clause shouldn't happen unless pre-London
+    if gas_price == U256::zero() {
+        return Err(GravityError::EthereumBadDataError(
+            "Latest block returned base fee per gas of zero".to_string(),
+        ));
+    }
+
+    let gas = min(GAS_LIMIT.into(), caller_balance.div_mod(gas_price).0);
+
+    Ok(GasCost { gas, gas_price })
+}
```

Comment:
Okay I think we should just use the Etherscan API here like we do in the Cellar Rebalancer and have user provide an Etherscan key.

Reply:
See https://github.com/PeggyJV/cellar_rebalancer_rs/blob/main/cellar_rebalancer/src/gas.rs

Reply:
Yeah, this was mirroring the web30 approach. It was also specifically for `call` rather than `send` which as far as I understand isn't meant to consume gas anyway, but we're still calculating these values. Based on your comment, it's OK to use the same gas calculation for both `call` and `send` then?